File: packages/afspec/afspec/lint.py
"""Spec linting: discover and validate specification packs.

Consolidates spec discovery, validation finding types, and the
lint runner into a single module.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from pathlib import Path

import afspec
from afspec.discovery import parse_spec_dir_name

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Finding types
# ---------------------------------------------------------------------------

SEVERITY_ERROR = "error"
SEVERITY_WARNING = "warning"
SEVERITY_HINT = "hint"

SEVERITY_ORDER = {SEVERITY_ERROR: 0, SEVERITY_WARNING: 1, SEVERITY_HINT: 2}


@dataclass(frozen=True)
class Finding:
    """A single validation finding."""

    spec_name: str
    file: str
    rule: str
    severity: str
    message: str
    line: int | None


def sort_findings(findings: list[Finding]) -> list[Finding]:
    return sorted(
        findings,
        key=lambda f: (f.spec_name, f.file, SEVERITY_ORDER.get(f.severity, 99)),
    )


def compute_exit_code(findings: list[Finding]) -> int:
    return 1 if any(f.severity == SEVERITY_ERROR for f in findings) else 0


# ---------------------------------------------------------------------------
# Spec discovery
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class SpecInfo:
    """Metadata about a discovered specification folder."""

    name: str
    prefix: int
    path: Path
    has_tasks: bool
    has_prd: bool


class LintError(Exception):
    """Raised when lint cannot proceed (e.g. missing specs directory)."""


def discover_specs(
    specs_dir: Path,
    filter_spec: str | None = None,
) -> list[SpecInfo]:
    """Discover spec folders with a requirements.json file present."""
    if not specs_dir.is_dir():
        raise LintError(f"No specifications found: '{specs_dir}' does not exist")

    specs: list[SpecInfo] = []
    found_candidates = False
    for entry in sorted(specs_dir.iterdir()):
        if not entry.is_dir():
            continue
        parsed = parse_spec_dir_name(entry.name)
        if parsed is None:
            continue

        found_candidates = True
        prefix, _ = parsed

        if not (entry / "requirements.json").is_file():
            logger.debug("Spec folder '%s' has no requirements.json, skipping", entry.name)
            continue

        has_tasks = (entry / "tasks.json").is_file()
        has_prd = (entry / "prd.md").is_file()

        if not has_tasks:
            logger.warning("Spec folder '%s' has no tasks.json", entry.name)

        specs.append(
            SpecInfo(name=entry.name, prefix=prefix, path=entry, has_tasks=has_tasks, has_prd=has_prd)
        )

    if not specs:
        if not found_candidates:
            raise LintError(f"No specifications found in '{specs_dir}'")
        return []

    specs.sort(key=lambda s: s.prefix)

    if filter_spec is not None:
        filtered = [s for s in specs if s.name == filter_spec]
        if not filtered:
            available = ", ".join(s.name for s in specs)
            raise LintError(f"Spec '{filter_spec}' not found. Available specs: {available}")
        return filtered

    return specs


# ---------------------------------------------------------------------------
# Lint runner
# ---------------------------------------------------------------------------

_KNOWN_SEVERITIES = {"error", "warning", "hint"}


def _map_afspec_findings(
    spec_name: str,
    errors: list[afspec.ValidationError],
) -> list[Finding]:
    findings: list[Finding] = []
    for ve in errors:
        raw_severity = getattr(ve, "severity", None)
        severity = raw_severity if raw_severity in _KNOWN_SEVERITIES else "error"
        findings.append(
            Finding(
                spec_name=spec_name,
                file=ve.file,
                rule=ve.rule,
                severity=severity,
                message=ve.message,
                line=getattr(ve, "line", None),
            )
        )
    return findings


def _validate_spec(spec: SpecInfo) -> list[Finding]:
    try:
        loaded = afspec.load_spec(spec.path)
        result = afspec.validate(loaded)
        return _map_afspec_findings(spec.name, result.errors)
    except Exception as exc:
        return [
            Finding(
                spec_name=spec.name,
                file=str(spec.path),
                rule="afspec-error",
                severity="error",
                message=str(exc),
                line=None,
            )
        ]


def _is_spec_implemented(spec: SpecInfo) -> bool:
    try:
        loaded = afspec.load_spec(spec.path)
        if not loaded.tasks or not loaded.tasks.task_groups:
            return False
        return all(all(st.state == afspec.SubtaskState.DONE for st in g.subtasks) for g in loaded.tasks.task_groups)
    except Exception:
        return False


@dataclass(frozen=True)
class LintResult:
    findings: list[Finding] = field(default_factory=list)
    exit_code: int = 0


def run_lint_specs(
    specs_dir: Path,
    *,
    lint_all: bool = False,
    progress_callback: None = None,
) -> LintResult:
    """Run spec linting and return structured results."""
    if not specs_dir.exists():
        raise LintError(f"Specs directory not found: {specs_dir}")

    if progress_callback is not None:
        progress_callback("Discovering specs...")
    try:
        discovered: list[SpecInfo] = discover_specs(specs_dir)
    except LintError:
        no_spec_finding = Finding(
            spec_name="(none)",
            file=str(specs_dir),
            rule="no-specs",
            severity="error",
            message=f"No specifications found in {specs_dir} directory",
            line=None,
        )
        return LintResult(findings=[no_spec_finding], exit_code=1)

    if not lint_all:
        filtered = [s for s in discovered if not _is_spec_implemented(s)]
        skipped = len(discovered) - len(filtered)
        if skipped > 0:
            logger.info("Skipping %d fully-implemented spec(s) (use --all to include)", skipped)
        if not filtered:
            return LintResult(findings=[], exit_code=0)
        discovered = filtered

    if progress_callback is not None:
        progress_callback(f"Validating {len(discovered)} spec(s)...")

    findings: list[Finding] = []
    for spec in discovered:
        findings.extend(_validate_spec(spec))

    findings = sort_findings(findings)
    exit_code = compute_exit_code(findings)
    return LintResult(findings=findings, exit_code=exit_code)

File: golang/lint.go
package afspec

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
)

// ---------------------------------------------------------------------------
// Lint module data structures (05-REQ-6)
// ---------------------------------------------------------------------------

// LintFinding represents a single lint diagnostic result with severity, rule,
// message, and location metadata.
type LintFinding struct {
	SpecName string
	File     string
	Rule     string
	Severity string // one of "error", "warning", "hint"
	Message  string
	Line     int // zero means absent
}

// LintSpecInfo contains metadata about a discovered spec folder eligible for
// linting.
type LintSpecInfo struct {
	Name     string
	Prefix   int
	Path     string
	HasTasks bool
	HasPRD   bool
}

// LintResult is the aggregate output of a lint run, containing all findings
// and an exit code.
type LintResult struct {
	Findings []LintFinding
	ExitCode int
}

// severityRank maps severity strings to sort ranks. Unknown severities
// are treated as "hint" (rank 2) per 05-REQ-6.E1.
func severityRank(s string) int {
	switch s {
	case "error":
		return 0
	case "warning":
		return 1
	case "hint":
		return 2
	default:
		return 2 // unknown severity treated as hint
	}
}

// ---------------------------------------------------------------------------
// DiscoverLintSpecs — spec folder discovery (05-REQ-7)
// ---------------------------------------------------------------------------

// DiscoverLintSpecs scans specsDir for subdirectories whose names are valid
// spec directory names, checks each for the presence of requirements.json,
// and returns a LintSpecInfo slice for all matching directories.
//
// When filterSpec is non-empty, only the entry whose Name matches exactly is
// returned; an error is returned if no match is found.
//
// HasTasks and HasPRD are set based on the presence of tasks.json and prd.md
// respectively.
func DiscoverLintSpecs(specsDir, filterSpec string) ([]LintSpecInfo, error) {
	entries, err := os.ReadDir(specsDir)
	if err != nil {
		return nil, fmt.Errorf("cannot read specs directory %q: %w", specsDir, err)
	}

	var infos []LintSpecInfo

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		name := entry.Name()

		// 05-REQ-7.1: use IsSpecDirName to check validity.
		if !IsSpecDirName(name) {
			continue
		}

		dir := filepath.Join(specsDir, name)

		// Only include directories that contain requirements.json.
		if _, err := os.Stat(filepath.Join(dir, "requirements.json")); err != nil {
			continue
		}

		// 05-REQ-7.E3: skip if ParseSpecDirName fails.
		prefixStr, specName, parseErr := ParseSpecDirName(name)
		if parseErr != nil {
			continue
		}

		prefix, convErr := strconv.Atoi(prefixStr)
		if convErr != nil {
			continue
		}

		hasTasks := fileExists(filepath.Join(dir, "tasks.json"))
		hasPRD := fileExists(filepath.Join(dir, "prd.md"))

		infos = append(infos, LintSpecInfo{
			Name:     specName,
			Prefix:   prefix,
			Path:     dir,
			HasTasks: hasTasks,
			HasPRD:   hasPRD,
		})
	}

	// 05-REQ-7.E2: error if no specs discovered at all.
	if len(infos) == 0 {
		return nil, fmt.Errorf("no specs found in %q", specsDir)
	}

	// 05-REQ-7.2: filter by name when filterSpec is non-empty.
	if filterSpec != "" {
		for _, info := range infos {
			if info.Name == filterSpec {
				return []LintSpecInfo{info}, nil
			}
		}
		return nil, fmt.Errorf("no spec with name %q found in %q", filterSpec, specsDir)
	}

	return infos, nil
}

// fileExists returns true if the path exists and is not a directory.
func fileExists(path string) bool {
	info, err := os.Stat(path)
	if err != nil {
		return false
	}
	return !info.IsDir()
}

// ---------------------------------------------------------------------------
// SortFindings and ComputeExitCode (05-REQ-8)
// ---------------------------------------------------------------------------

// SortFindings returns a new slice sorted primarily by SpecName ascending,
// then by File ascending, then by Severity in the order error < warning < hint.
// The input slice is not modified.
func SortFindings(findings []LintFinding) []LintFinding {
	sorted := make([]LintFinding, len(findings))
	copy(sorted, findings)

	sort.SliceStable(sorted, func(i, j int) bool {
		a, b := sorted[i], sorted[j]
		if a.SpecName != b.SpecName {
			return a.SpecName < b.SpecName
		}
		if a.File != b.File {
			return a.File < b.File
		}
		return severityRank(a.Severity) < severityRank(b.Severity)
	})

	return sorted
}

// ComputeExitCode returns 1 if any finding has Severity "error", 0 otherwise.
func ComputeExitCode(findings []LintFinding) int {
	for _, f := range findings {
		if f.Severity == "error" {
			return 1
		}
	}
	return 0
}

// ---------------------------------------------------------------------------
// RunLintSpecs — full lint execution (05-REQ-9)
// ---------------------------------------------------------------------------

// isFullyImplemented checks whether a spec's tasks.json exists and all
// subtasks across all groups have state 'done' or 'dropped'. Returns false
// if tasks.json cannot be read, parsed, or has any subtask in another state.
func isFullyImplemented(info LintSpecInfo) bool {
	if !info.HasTasks {
		return false
	}

	tasksPath := filepath.Join(info.Path, "tasks.json")
	data, err := os.ReadFile(tasksPath)
	if err != nil {
		return false
	}

	var tasks TasksV1Json
	if err := json.Unmarshal(data, &tasks); err != nil {
		return false
	}

	for _, group := range tasks.TaskGroups {
		for _, sub := range group.Subtasks {
			if sub.State != SubtaskStateDone && sub.State != SubtaskStateDropped {
				return false
			}
		}
	}

	return true
}

// validationEntryToFinding converts a ValidationEntry into a LintFinding for
// the named spec. The severity is determined by the entry's category: entries
// with category "warning" get severity "warning"; all others get "error".
func validationEntryToFinding(specName string, entry ValidationEntry) LintFinding {
	severity := "error"
	if entry.Category == "warning" {
		severity = "warning"
	}

	rule := entry.Check
	if rule == "" {
		rule = entry.Category
	}

	return LintFinding{
		SpecName: specName,
		File:     entry.Artifact,
		Rule:     rule,
		Severity: severity,
		Message:  entry.Message,
	}
}

// RunLintSpecs scans specsDir for spec folders, loads and validates each,
// and returns a LintResult with sorted findings and computed exit code.
// When lintAll is false, fully-implemented specs (all subtasks in state
// 'done' or 'dropped') are skipped.
func RunLintSpecs(specsDir string, lintAll bool) (LintResult, error) {
	infos, err := DiscoverLintSpecs(specsDir, "")
	if err != nil {
		return LintResult{}, err
	}

	var allFindings []LintFinding

	for _, info := range infos {
		// Skip fully-implemented specs when lintAll is false.
		if !lintAll && isFullyImplemented(info) {
			continue
		}

		// Load the spec from disk.
		spec, loadErr := LoadSpec(info.Path)
		if loadErr != nil {
			// 05-REQ-9.E2 / 05-ERR-8: record a load-failure finding and
			// continue processing remaining specs.
			allFindings = append(allFindings, LintFinding{
				SpecName: info.Name,
				File:     "requirements.json",
				Rule:     "load-failure",
				Severity: "error",
				Message:  loadErr.Error(),
			})
			continue
		}

		// Validate the loaded spec.
		result := spec.Validate()

		// Convert validation errors to LintFinding entries.
		for _, e := range result.Errors {
			allFindings = append(allFindings, validationEntryToFinding(info.Name, e))
		}

		// Convert validation warnings to LintFinding entries.
		for _, w := range result.Warnings {
			allFindings = append(allFindings, validationEntryToFinding(info.Name, w))
		}
	}

	sorted := SortFindings(allFindings)
	exitCode := ComputeExitCode(sorted)

	return LintResult{
		Findings: sorted,
		ExitCode: exitCode,
	}, nil
}
